### DISCUSSION TITLE
Refresh data reduction on zoom

### DISCUSSION THREAD

**Discussion by Kusefiru** [GSI-CS-CO/chart-fx#393] [Ideas]
Hi,

Firstly, thank you for your work on Chart-Fx.

We are currently building an application using the library to display multiple huge datasets.
The library works well for 5 to 10 million points, however our clients would like to display even larger datas, going up to 50 to 100 million points. While it is possible to display such datasets, navigating them become difficult as the application slows down considerably.
My first idea to improve on the performances was to use one of your pre-included data reduction algorithm. Considering the fuzziness of our datasets, only the MaxDataReducer reducer has been able to reduce them to a decent amount of points, despite losing a lot of informations on the way.
From my undertanding on how the library work, the reduction is applied before displaying the data and as such, zooming on the data only zoom on the "reduced" dataset. What we would like would be to recompute the reduced dataset based on the current viewable window. Is there any plan to include such a feature to the library ?

I am not experienced much in Java so I may have missed something.

**Comment by wirew0rm** [comment]
Hey,

thank you for your interest in chart-fx.

Actually the reduction step is performed for every render() call, so there is no caching going on and its only purpose is not to overwhelm the rendering pipeline and also to correctly visualize errors for dense datasets.

In your application it seems, that it is infeasable to process the whole dataset for each render call, so you would need some caching for data at different zoom levels.
You could wrap your DataSet in a custom DataSet(DataSet rawData, AxisRange range) which depending on the range of your xAxis returns data at different zoom levels and keeps a cache for the zoom levels.

This should not be too complicated and I think it can be implemented in user code. At the moment we do not have the usecases/resources to add this to the library, but if you find a good solution and want to contribute it we won't turn down a pull request ;-)

Hope this helps!

**Comment by Kusefiru** [comment]
Thank you for your reply !

I understand your idea, and it seems doable.
I'll look into it and submit a pull request if I come up with something good.
